feat(embind): add support for type-based overloading

bugra9 opened the pull request:
```cpp
 class A {};
 class B : public A {};
 class C {};

 class HasOverloadedMethods {
 public:
     HasOverloadedMethods(std::string i); // javascript type of parameter: string
     HasOverloadedMethods(int i); // javascript type of parameter: number

     void foo();
     void foo(std::string i); // javascript type of parameter: string
     void foo(short i); // javascript type of parameter: number
     void foo(A i); // javascript type of parameter: A
     void foo(std::shared_ptr<C> i); // javascript type of parameter: C
     void foo(int i, int j); // javascript types of parameters: number, number
     void foo(double i); // javascript type of parameter: number
     void foo(float f) const; // javascript type of parameter: number
     void foo(B i); // javascript type of parameter: B (derived from A)
     
     static void staticFoo(std::string i); // javascript type of parameter: string
     static void staticFoo(int i); // javascript type of parameter: number
};

 // Binding code
 EMSCRIPTEN_BINDING(overloads) {
     class_<HasOverloadedMethods>("HasOverloadedMethods")
         .constructor<std::string>()
         .constructor<int>()
         // .smart_ptr_constructor("HasOverloadedMethods", &std::make_shared<HasOverloadedMethods, std::string>)
         // .smart_ptr_constructor("HasOverloadedMethods", &std::make_shared<HasOverloadedMethods, int>)
         .function("foo", select_overload<void()>(&HasOverloadedMethods::foo))
         .function("foo", select_overload<void(std::string)>(&HasOverloadedMethods::foo))
         .function("foo", select_overload<void(short)>(&HasOverloadedMethods::foo))
         .function("foo", select_overload<void(A)>(&HasOverloadedMethods::foo))
         .function("foo", select_overload<void(std::shared_ptr<C>)>(&HasOverloadedMethods::foo))
         .function("foo", select_overload<void(int, int)>(&HasOverloadedMethods::foo))
         .function("foo_double", select_overload<void(double)>(&HasOverloadedMethods::foo))
         .function("foo_float", select_overload<void(float)const>(&HasOverloadedMethods::foo))
         .function("foo_B", select_overload<void(B)>(&HasOverloadedMethods::foo))
         .class_function("staticFoo", select_overload<void(std::string)>(&HasOverloadedMethods::staticFoo))
         .class_function("staticFoo", select_overload<void(int)>(&HasOverloadedMethods::staticFoo))
         ;
 }
```

- [x] Functions can be overloaded using parameter js types.
- [x] Class methods can be overloaded using parameter js types.
- [x] Static class methods can be overloaded using parameter js types.
- [x] Class constructors can be overloaded using parameter js types. (constructor and smart_ptr_constructor)
- [x] Add tests.
- [x] Update the documentation.
- [x] Run `test/runner other.test_embind`.
- [x] Run `EMTEST_SKIP_V8=1 test/runner core0 skip:core0.test_time`.

brendandahl commented on src/embind/embind.js:
```diff
@@ -97,23 +97,93 @@ var LibraryEmbind = {
+        break;
+      } else if (
+        tempObject.constructor.name === field
+        || 'shared_ptr<' + tempObject.constructor.name + '>' === field
```
This seems brittle. Is there some other way we can handle pointers? Maybe, once the instanceof works we can get a ref to the class that the pointer points to.

bugra9 replied:
I deleted instancesCheck function and now i use `instanceof` instead.

brendandahl reviewed (changes requested):
For current consumers of the overload feature this is going to make things much slower, since there's a lot more work that will happen on every call to an overloaded function. Some thoughts on how we could improve this:

- First do a check for argument length, if there's only one method that matches call it. Otherwise, do the expensive type checking.
- There's a lot of split/joining and string matching. I think not relying on strings and keeping things split will be more efficient (see some comments inline).

Last, please update the [documentation](https://github.com/emscripten-core/emscripten/blob/main/site/source/docs/porting/connecting_cpp_and_javascript/embind.rst).

bugra9 commented on src/embind/embind.js:
```diff
@@ -344,7 +417,7 @@ var LibraryEmbind = {
         }
         for (var i = 0; i < myTypes.length; ++i) {
-            registerType(myTypes[i], myTypeConverters[i]);
+            registerType(myTypes[i], myTypeConverters[i], { ignoreDuplicateRegistrations: true });
```
I needed it to use smart_ptr_constructor function multiple. (smart_ptr_constructor function call smart_ptr function in bind.h.) I updated this code to pass only for the smart_ptr_constructor.

bugra9 replied:
I deleted it.

bugra9 commented on src/embind/embind.js:
```diff
@@ -97,23 +97,93 @@ var LibraryEmbind = {
+        }
+        break;
+      } else if (
+        tempObject.constructor.name === field
```
I deleted instancesCheck function and now i use `instanceof` instead.

bugra9 commented on src/embind/embind.js:
```diff
@@ -97,23 +97,93 @@ var LibraryEmbind = {
       // Inject an overload resolver function that routes to the appropriate overload based on the number of arguments.
       proto[methodName] = function() {
+        var args = Array.prototype.slice.call(arguments);
+        var keys = Object.keys(proto[methodName].overloadTable).map(k => k.split(', '));
```
done

bugra9 commented on src/embind/embind.js:
```diff
@@ -122,49 +192,52 @@ var LibraryEmbind = {
     // If there's an overload table for this symbol, replace the symbol in the overload table instead.
-    if (undefined !== Module[name].overloadTable && undefined !== numArguments) {
-      Module[name].overloadTable[numArguments] = value;
+    if (undefined !== Module[name].overloadTable && undefined !== signature) {
```
done

bugra9 commented:
Done. I refactored all my code performance-oriented and updated the documentation.

> For current consumers of the overload feature this is going to make things much slower, since there's a lot more work that will happen on every call to an overloaded function. Some thoughts on how we could improve this:
> 
> * First do a check for argument length, if there's only one method that matches call it. Otherwise, do the expensive type checking.
> * There's a lot of split/joining and string matching. I think not relying on strings and keeping things split will be more efficient (see some comments inline).
> 
> Last, please update the [documentation](https://github.com/emscripten-core/emscripten/blob/main/site/source/docs/porting/connecting_cpp_and_javascript/embind.rst).

bugra9 commented:
Hi @brendandahl, do you have any suggestions?

bugra9 committed:
feat(embind): add support for type-based overloading

erenkucuker commented:
Hi, is there any plan to merge it ?
